src/GameEngine/GameEngine.cpp

```diff
@@ -0,0 +1,161 @@
+//
+// Created by ker10 on 2019-10-23.
+//
+
+#include "GameEngine.h"
+#include <windows.h>
+#include "../Map/Map.h"
+#include "../MapLoader/MapLoader.h"
+#include "../Terminal/Terminal.h"
+#include "../exceptions.h"
+
+using std::to_string;
+
+namespace GameEngine {
+    GameEngine *GameEngine::game_engine_instance = nullptr;
+
+    GameEngine *GameEngine::instance() {
+        if (!game_engine_instance) game_engine_instance = new GameEngine;
+        return game_engine_instance;
+    }
+
+    /*
+     * get_available_map list the ressources/maps/ directory and removes the .map from the file
+     * It uses windows.h :: FindNextFile() to traverse the directory.
+     *
+     * TODO: make it cross-platform with mac since some team member uses it
+     *
+     * https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-findnextfilea
+     */
+    vector<string> GameEngine::get_available_map() {
+        vector<string> results;
+
+        string search_path = "ressources/maps/*.*";
+        WIN32_FIND_DATA fd;
+        HANDLE h_find = FindFirstFile(search_path.c_str(), &fd);
+        if(h_find != INVALID_HANDLE_VALUE) {
+            do {
+                if(! (fd.dwFileAttributes & FILE_ATTRIBUTE_DIRECTORY) ) {
+                    string map_name = fd.cFileName;
+                    size_t extension = map_name.find(".map");
+                    map_name = map_name.substr(0, extension);
+                    results.emplace_back(map_name);
+                }
+            } while( FindNextFile(h_find, &fd)) ;
+            FindClose(h_find);
+        }
+
+        return results;
+    }
+
+    void GameEngine::select_map() {
+        MapLoader::MapLoader mapLoader;
+        Terminal::print("Select the map you want to play in.");
+        vector<string> maps = get_available_map();
+        int map_selection = Terminal::print_select(maps);
+        try {
+            map = mapLoader.load(map_selection)->build();
+        } catch (IOException &e ) {
+            string error = e.what(maps[map_selection]);
+            Terminal::error(error);
+        } catch(exception &e){
+            string error = e.what();
+            Terminal::error(error);
+        }
+        Terminal::debug("Map defined in GameEngine");
+    }
+
+    void GameEngine::create_deck() {
```

Comment:
Should the deck be shuffled as well?

Reply:
Didnt meant to make that resolved, but yeah it could be shuffled here. I don’t know if it makes a difference later on